lib/active_model/serializer.rb

```diff
@@ -112,10 +112,24 @@ def self.digest_caller_file(caller_line)
       Digest::MD5.hexdigest(serializer_file_contents)
     end
 
+    def self.serializer_lookup_chain_for(klass)
+      chain = []
+
+      resource_class_name = klass.name.demodulize
+      resource_namespace = klass.name.deconstantize
+      serializer_class_name = "#{resource_class_name}Serializer"
+
+      chain.push("#{name}::#{serializer_class_name}") if self != ActiveModel::Serializer
+      chain.push("#{resource_namespace}::#{serializer_class_name}")
+      chain.push(serializer_class_name)
+
+      chain
+    end
+
     def self.get_serializer_for(klass)
       serializers_cache.fetch_or_store(klass) do
-        serializer_class_name = "#{klass.name}Serializer"
-        serializer_class = serializer_class_name.safe_constantize
+        # NOTE(beauby): When we drop 1.9.3 support we can lazify the map for perfs.
+        serializer_class = serializer_lookup_chain_for(klass).map(&:safe_constantize).find { |x| x }
```

Comment:
the chain is less efficient than my suggestion and isn't a very revealing method name.  Lemme try to do it:

```ruby
    SerializerForResource = Struct.new(:lookup_serializer, :resource_klass) do
      def lookup
        serializer_class = nil
        potential_serializer_class_names.each do |serializer_class_name|
          serializer_class = serializer_class_name.safe_constantize
          break if serializer_class
        end

        if serializer_class
          serializer_class
        elsif klass.superclass
          lookup_serializer.get_serializer_for(resource_klass.superclass)
        end
      end

      # @example
      #   lookup_serializer = AdminSerializer
      #   resource_klass = Admin::User
      #   SerializerForResource.new(lookup_serializer, resource_klass).lookup
      #   #=> %w[AdminSerializer::UserSerializer Admin::UserSerializer UserSerializer]
      def potential_serializer_class_names
        chain = []
        # Given Admin::User
        resource_class_name = klass.name.demodulize    # User
        resource_namespace = klass.name.deconstantize # Admin
        serializer_class_name = "#{resource_class_name}Serializer"

        chain.push("#{name}::#{serializer_class_name}") if lookup_serializer != ActiveModel::Serializer
        chain.push("#{resource_namespace}::#{serializer_class_name}")
        chain.push(serializer_class_name)

        chain
      end
    end
    def self.get_serializer_for(resource_klass)
      serializers_cache.fetch_or_store(resource_klass) do
        SerializerForResource.new(self, resource_klass).lookup
      end
    end
```